clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java

```diff
@@ -228,7 +229,8 @@ private void poll(long timeout, long now, boolean executeDelayedTasks) {
         trySend(now);
 
         // fail all requests that couldn't be sent
-        failUnsentRequests();
+        if (failUnsent)
+            failUnsentRequests();
```

Comment:
Sounds good. I'm also ok with a more incremental change if it ends up being more complex than I suggested.